src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -432,16 +440,30 @@ private void AddDirectoryWatchUnlocked(WatchedDirectory? parent, string director
                 // asked for subdirectories to be included.
                 if (isNewDirectory && _includeSubdirectories)
                 {
-                    // This method is recursive.  If we expect to see hierarchies
-                    // so deep that it would cause us to overflow the stack, we could
-                    // consider using an explicit stack object rather than recursion.
-                    // This is unlikely, however, given typical directory names
-                    // and max path limits.
-                    foreach (string subDir in Directory.EnumerateDirectories(fullPath))
+                    try
                     {
-                        AddDirectoryWatchUnlocked(directoryEntry, System.IO.Path.GetFileName(subDir));
-                        // AddDirectoryWatchUnlocked will add the new directory to
-                        // this.Children, so we don't have to / shouldn't also do it here.
+                        // This method is recursive.  If we expect to see hierarchies
+                        // so deep that it would cause us to overflow the stack, we could
+                        // consider using an explicit stack object rather than recursion.
+                        // This is unlikely, however, given typical directory names
+                        // and max path limits.
+                        foreach (string subDir in Directory.EnumerateDirectories(fullPath))
+                        {
+                            AddDirectoryWatchUnlocked(directoryEntry, System.IO.Path.GetFileName(subDir));
+                            // AddDirectoryWatchUnlocked will add the new directory to
+                            // this.Children, so we don't have to / shouldn't also do it here.
+                        }
+                    }
+                    catch (DirectoryNotFoundException)
+                    { } // The child directory was removed.
+                    catch (IOException ex) when (ex.HResult == Interop.Error.ENOTDIR.Info().RawErrno)
+                    { } // The child directory was replaced by a file.
+                    catch (Exception ex)
+                    {
+                        if (_weakWatcher.TryGetTarget(out FileSystemWatcher? watcher))
```

Comment:
In the reported issue `Directory.EnumerateDirectories` failed with `DirectoryNotFoundException` so this particular catch is for this method.

I assume that you added `ENOTDIR` because you know that it's another error that could potentially occur?

But how about the `catch (Exception ex)`? You have done that just to be extra safe? I can see that we are doing that in other places here. I am not 100% sure if all our unit tests verify that `OnError` was not invoked. Can we add `Debug.Fail(ex.ToString());` so when we introduce a bug here the CI is going to get red?

```suggestion
                    {
                        Debug.Fail(ex.ToString());

                        if (_weakWatcher.TryGetTarget(out FileSystemWatcher? watcher))
```

Reply:
> I assume that you added ENOTDIR because you know that it's another error that could potentially occur?

`inotify` is not recursive. When we get a notification that a new directory is added, we need to recurse it ourselves. By the time we start recursing it, the directory may be gone (`DirectoryNotFoundException`) or replaced by a file (`ENOTDIR`). In those two cases, we pretend the directory was never there.

If there are other errors, we report them to the user, similar to what already happens for `inotify_add_watch` errors.

> Can we add Debug.Fail(ex.ToString()); so when we introduce a bug here the CI is going to get red?

It isn't there for `inotify_add_watch` errors, so I did the same thing here.